### ISSUE TITLE
Akka http client getting stuck

### ISSUE THREAD

**Issue by jmnght** [akka/akka-http#2529]
The pool connections in akka http client (Java DSL) is getting stuck in WaitingForResponse state and starts applying backpressure making the whole application unusable in heavy load when the ```singleRequest``` endpoint is slow to respond.

The pool settings is default as below:
Pool conn timeouts getIdleTimeout: 30 seconds getMaxConnections:
 4 getMaxOpenRequests: 32 getMinConnections:0 getMaxRetries: 5 getConnectingTimeout: 10 sec
onds getIdleTimeout: 1 minute

And it's getting stuck here:
```DEBUG[mylog-dispatcher-blocking] a.h.i.e.c.PoolGateway - [3 (WaitingForResponse)] After event [onRequestDispatched] State change [PushingRequestToConnection] -> [WaitingForResponse]```

Is the connections in default pool somehow blocking?

The documentation for slow response points to Akka http client Connection Level client side API while that page kind of discourages its use as it ends up creating a new connection for every request, which is expensive.

Not sure if this is a bug or missing documentation of how to properly configure http client in Java DSL for heavy load on slow responding requests.
 
Thanks in advance for your inputs.

**Comment by jrudolph** [comment]
Thanks for the report, @jmnght. `WaitingForResponse` actually just means that, one of the pool connections is waiting for a response. With `max-connections = 4` you will get 4 connections per host. If all those are waiting for a response no new requests will be sent. We currently don't have a dedicated request timeout, however, after a while the `idle-timeout` will kick in and fail those requests. Requests are then retried (if eligible), so if at the front of the queue to this host there are four slow or hanging requests they might keep the connection pool for that host busy for `(max-retries + 1) * idle-timeout = 6 minutes` in your case.

> Is the connections in default pool somehow blocking?

Not in the sense of blocking a thread but as mentioned above connections per host are a limited resource and if all connections are busy no new requests will be handled.

What can you do? You can either decrease the `idle-timeout` so that slow/hanging requests are cancelled earlier. Or, you can increase `max-connections`, the number of connections per host. If you have rather slow requests than that would make sense.

Btw. to debug what's going out or coming in on the wire, you can also use `akka.http.client.log-unencrypted-network-bytes = 1000` or so to see if a response comes in.

**Comment by jmnght** [comment]
Thank you for the details @jrudolph
The weird thing is I don't see connections being closed after 6 minutes in this case. Shouldn't there be another log that says returning connection after idle time out or retry? The application just hangs infinitely after this..

In my case the application exposes few REST endpoints via akka http server and for each request it uses the akka http client to call another external service, get the response back, materializes response, apply some logic /processing and return response. If I don't get anything back from external service, as I mentioned the whole application including http server hangs.
I was expecting some debug logs at least when the backpressure kicks in, but don't see anywhere.

Also, I'm thinking of making the idle time out to 5 seconds with 0 retry and max connection of 1000 or so. I will test it out and see report what I find.

**Comment by jrudolph** [comment]
>  I was expecting some debug logs at least when the backpressure kicks in, but don't see anywhere.

There should be some log messages definitely when the `idle-timeout` kicks in. When the problem persists please post complete configuration and relevant DEBUG logs with `akka.http.client.log-unencrypted-network-bytes = 1000`.

**Comment by jmnght** [comment]
@jrudolph This seems to happen again in heavier loads even with 5 seconds and 0 retries.
My Http Client class looks like below:

```
public class MyAkkaHttpClient {

	private final LoggingAdapter logger = Logging.getLogger(MyAkkaConfig.INSTANCE.getSystem(), this);
	private final Http http;
	private final ObjectMapper mapper;
	private static final MyAkkaHttpClient instance = new MyAkkaHttpClient();
	private final ConnectionPoolSettings poolSettingsWithHttpsProxy;

	private HotelbedsAkkaHttpClient() {
		ActorSystem system = MyAkkaConfig.INSTANCE.getSystem();
		http = Http.get(system);
		
		mapper = new ObjectMapper();
		
		ClientTransport proxy = ClientTransport
				.httpsProxy(InetSocketAddress.createUnresolved(MyHttpConfig.getProxyHost(), MyHttpConfig.getProxyPort()));
		poolSettingsWithHttpsProxy = ConnectionPoolSettings.create(system)
				.withConnectionSettings(ClientConnectionSettings.create(system).withTransport(proxy));
	}
	
	public static HttpClient getInstance() { return instance; }

	public MyResponse doPost(MyRequest myReq) {

		RequestEntity entity = HttpEntities.create(ContentTypes.APPLICATION_JSON, mapper.writeValueAsString(myReq));
		
		HttpRequest httReq =  HttpRequest.POST(MyHttpConfig.getUrl())
				.addHeaders(Arrays.asList(
						HttpHeader.parse("Accept", "application/json"),
						HttpHeader.parse("Accept-Encoding", "application/gzip"))
				.withEntity(entity);
		
		return http.singleRequest(httReq, http.defaultClientHttpsContext(), poolSettingsWithHttpsProxy, logger)
				.thenCompose(response -> {
					if (response.status().isSuccess()) {
			
						return Jackson.unmarshaller(mapper, MyResponse.class).unmarshal(response.entity(), ExecutionContext.global(),
								AkkaConfig.INSTANCE.getMaterializer());
					} else {
			
						return response.entity()
		               	 		.getDataBytes()
		                		.runWith(Sink.<ByteString>head(), AkkaConfig.INSTANCE.getMaterializer())
		                		.toCompletableFuture()
								.thenApply((responseBody) -> {
				
									MyResponse rs = new MyResponse();
									rs.setError(new Error(response.status().toString(), responseBody.utf8String()));
									return rs;
								});
					}
		}).toCompletableFuture()
			.exceptionally(t -> defaultMyResponse(t)).join();
	}
}
```
Do you see any issues?
Looks like the connection is stuck here:
```
2019-05-17 12:07:33.250 DEBUG[HttsServer-logging-blocking-dispatcher-5] a.h.i.e.c.PoolGateway - [2 (WaitingForEndOfResponseEntity)] After event [onResponseEntitySubscribed] State change [WaitingForResponseEntitySubscription] -> [WaitingForEndOfResponseEntity]
2019-05-17 12:07:33.942 DEBUG[HttsServer-logging-blocking-dispatcher-5] a.h.i.e.c.PoolGateway - [3 (WaitingForResponse)] After event [onRequestDispatched] State change [PushingRequestToConnection] -> [WaitingForResponse]
```
I will PM you the full logs.. Thanks!

**Comment by jrudolph** [comment]
The logs look like nothing was received from the other side. Since the idle timeout doesn't seem to kick in, one potential explanation could be that you are suffering from thread starvation? I.e. some other things in your ActorSystem are blocking threads so that akka-http cannot make any progress any more? Can you check with `jstack <pid>` what the dispatcher threads are doing in such a situation?

**Comment by jmnght** [comment]
@jrudolph Every incoming request to my akka http server is calling the ```doPost``` above and from the ```jstack```, it looks like the incoming requests are blocked by the ```join()``` at the end of the ```doPost``` method. Does that sound right? 
Also, do you see any other issues with above code on eyeballing it?

**Comment by jrudolph** [comment]
Ah, yes, I didn't look to closely before. That's the problem. If you block there, then it will prevent other work from being executed.

Can you change your code in a way that the response is handled asynchronously, i.e. return `CompletableFuture<MyResponse>` from `doPost`? In calling code, you also must not call `join` but use one of the `CompletableFuture` combinators to continue with the result, e.g. `thenApply`.

**Comment by jmnght** [comment]
@jrudolph Got a little bit stuck trying to change it to returning ```CompletableFuture```.
This call was part of a Graph which now looks like this:
```
public Graph<FlowShape<ClassA, CompletionStage<ClassD>>, NotUsed> process() {
    return ...;
}
 ```
Trying to materialize it in route results in: 
```CompletionStage<CompletionStage<ClassD>> response = source.via(process()).runWith(Sink.head(), materializer)
```

**Comment by jrudolph** [comment]
You can use the mapAsync stream operator to unpack one level of futures.

**Comment by jmnght** [comment]
@jrudolph That worked perfectly and fixed the hanging issue, but it opened up another issue which is with BufferOverflow:
```
akka.stream.BufferOverflowException: Exceeded configured max-open-requests value of [32]. This means that the request queue of this pool (HostConnectionPoolSetup(api.hbs*.com,443,ConnectionPoolSetup(ConnectionPoolSettings(4,0,0,32,1,100 milliseconds,2 minutes,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/10.1.7),10 seconds,1 minute,512,Some(1000),WebSocketSettings(<function0>,ping,Duration.Inf,akka.http.impl.settings.WebSocketSettingsImpl$$$Lambda$396/1623060829@24dd186),List(),ParserSettings(2048,16,64,64,8192,64,8388608,8388608,256,1048576,Strict,RFC6265,true,Set(),Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,true,akka.util.ConstantFun$$$Lambda$219/1097897234@25caeccd,akka.util.ConstantFun$$$Lambda$219/1097897234@25caeccd,akka.util.ConstantFun$$$Lambda$220/1107530534@2ddf44d),None,HttpsProxyTransport(nw-prx-v01.dqs.pcln.com:8080,TCPTransport,None)),New,1 second),akka.http.scaladsl.HttpsConnectionContext@618f56,akka.event.BusLogging@799b658a))) has completely filled up because the pool currently does not process requests fast enough to handle the incoming request load. Please retry the request later. See http://doc.akka.io/docs/akka-http/current/scala/http/client-side/pool-overflow.html for more information.
```

**Comment by jrudolph** [comment]
> But there is also an issue raised that says not to use that: #738

Using `singleRequest` (with different settings) would indeed be preferred if you don't provide the requests as a stream in the first place.

> I create 2 different `Http` with different `ConnectionPoolSettings`, I'm thinking that would better solve isolating long running requests from others, right?

You mean, calling `Http.singleRequest` with different settings? Yes, that should work.

> Will the `max-open-requests` be shared between the 2 pools?

No, so you are good there.

**Comment by jmnght** [comment]
Thank you @jrudolph greatly appreciate all your help. So far the application is holding up fine in production as POC. We will be ramping up the traffic in the next few weeks and will report back on any findings/experience.
